core/templates/services/exploration-improvements-task-registrar.service.ts

```diff
@@ -0,0 +1,424 @@
+ * exploration.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { merge } from 'd3-array';
+import { sortBy } from 'lodash';
+
+import { AnswerStats } from 'domain/exploration/AnswerStatsObjectFactory';
+import {
+  CyclicStateTransitionsPlaythrough,
+  EarlyQuitPlaythrough,
+  MultipleIncorrectSubmissionsPlaythrough,
+  Playthrough,
+} from 'domain/statistics/PlaythroughObjectFactory';
+import { ExplorationStats } from
+  'domain/statistics/ExplorationStatsObjectFactory';
+import {
+  ExplorationTask,
+  ExplorationTaskObjectFactory,
+  ExplorationTaskType,
+} from 'domain/improvements/ExplorationTaskObjectFactory';
+import { HighBounceRateTask } from
+  'domain/improvements/HighBounceRateTaskObjectFactory';
+import { ITaskEntryPayloadDict } from
+  'domain/improvements/TaskEntryObjectFactory';
+import { ImprovementsConstants } from
+  'domain/improvements/improvements.constants';
+import { IneffectiveFeedbackLoopTask } from
+  'domain/improvements/IneffectiveFeedbackLoopTaskObjectFactory';
+import { NeedsGuidingResponsesTask } from
+  'domain/improvements/NeedsGuidingResponsesTaskObjectFactory';
+import { States } from 'domain/exploration/StatesObjectFactory';
+import { SuccessiveIncorrectAnswersTask } from
+  'domain/improvements/SuccessiveIncorrectAnswersTaskObjectFactory';
+
+type HbrTask = HighBounceRateTask;
+type IflTask = IneffectiveFeedbackLoopTask;
+type NgrTask = NeedsGuidingResponsesTask;
+type SiaTask = SuccessiveIncorrectAnswersTask;
+type CstPlaythrough = CyclicStateTransitionsPlaythrough;
+type EqPlaythrough = EarlyQuitPlaythrough;
+type MisPlaythrough = MultipleIncorrectSubmissionsPlaythrough;
+
+class TaskRegistry {
+  public readonly hbrTask: HbrTask;
+  public readonly iflTask: IflTask;
+  public readonly ngrTask: NgrTask;
+  public readonly siaTask: SiaTask;
+
+  public readonly initialHbrTaskStatus: string;
+  public readonly ngrTaskWasInitiallyOpen: boolean;
+
+  constructor(tasks: ExplorationTask[]) {
+    const [hbrTask, iflTask, ngrTask, siaTask] = sortBy(tasks, t => t.taskType);
+    this.hbrTask = <HighBounceRateTask> hbrTask;
+    this.initialHbrTaskStatus = hbrTask.getStatus();
+    this.iflTask = <IneffectiveFeedbackLoopTask> iflTask;
+    this.ngrTask = <NeedsGuidingResponsesTask> ngrTask;
+    this.ngrTaskWasInitiallyOpen = ngrTask.isOpen();
+    this.siaTask = <SuccessiveIncorrectAnswersTask> siaTask;
+  }
+
+  public forEach(fn: (t: ExplorationTask) => void): void {
+    fn(this.hbrTask);
+    fn(this.iflTask);
+    fn(this.ngrTask);
+    fn(this.siaTask);
+  }
+
+  public map(fn: (t: ExplorationTask) => ExplorationTask): TaskRegistry {
+    return new TaskRegistry([
+      fn(this.hbrTask), fn(this.iflTask), fn(this.ngrTask), fn(this.siaTask),
+    ]);
+  }
+
+  public didHbrTaskStatusChange(): boolean {
+    return this.initialHbrTaskStatus !== this.hbrTask.getStatus();
+  }
+}
+
+class StatsRegistry {
```

Comment:
This (and possibly other things) need comprehensive documentation on what they are.

"Registry" is a very general term, to the point where it doesn't actually convey that much meaning, so some elucidation in the form of comments would be helpful. There may be other places which need this too (particularly where the "why" of the function is not obvious from a casual reading).